### PR TITLE
feat: find Python files in Rust

### PR DESCRIPTION
**Pull request by mkniewallner**
**PR Checklist**

-   [x] A description of the changes is added to the description of this PR.
-   [ ] If there is a related issue, make sure it is linked to this PR.
-   [ ] If you've fixed a bug or added code that should be tested, add tests!
-   [ ] Documentation in `docs` is updated

**Description of changes**

Leverage Rust to find Python files by using https://crates.io/crates/ignore. The crate provides similar options to https://pypi.org/project/pathspec/ in order to handle `.gitignore` while also handling files walking. For the exclusion, the `regex` crate that we already depend on is used, to match the behaviour we have with the Python implementation. At some point, although this would be a breaking change, I would really like to move to globsets instead of regexes.

In terms of performance, testing the changes on a repository with ~7k files shows a 27% speedup, dropping from 1.43s to 1.04s.

### CODE REVIEW

File: python/deptry/core.py [hunk 1]
```diff
@@ -14,7 +14,7 @@
 from deptry.exceptions import IncorrectDependencyFormatError, UnsupportedPythonVersionError
 from deptry.imports.extract import get_imported_modules_from_list_of_files
 from deptry.module import ModuleBuilder, ModuleLocations
-from deptry.python_file_finder import PythonFileFinder
+from deptry.python_file_finder import get_all_python_files_in
```
Comment (fpgmaas) [inline python/deptry/core.py:17]: Is it an idea to directly use the function that is defined in Rust here, and remove `python_file_finder.py` completely?

```suggestion
from deptry.rust import find_python_files
```

**Comment by fpgmaas** [comment]
Looks good! I see that the verbose scanning has become a bit more verbose:

```
Collecting Python files to scan...
built glob set; 0 literals, 0 basenames, 0 extensions, 0 prefixes, 0 suffixes, 2 required extensions, 0 regexes
glob converted to regex: Glob { glob: "**/*.py[cod]", re: "(?-u)^(?:/?|.*/)[^/]*\\.py[cod]$", opts: GlobOptions { case_insensitive: false, literal_separator: true, backslash_escape: true, empty_alternates: false }, tokens: Tokens([RecursivePrefix, ZeroOrMore, Literal('.'), Literal('p'), Literal('y'), Class { negated: false, ranges: [('c', 'c'), ('o', 'o'), ('d', 'd')] }]) }
glob converted to regex: Glob { glob: "**/.coverage.*", re: "(?-u)^(?:/?|.*/)\\.coverage\\.[^/]*$", opts: GlobOptions { case_insensitive: false, literal_separator: true, backslash_escape: true, empty_alternates: false }, tokens: Tokens([RecursivePrefix, Literal('.'), Literal('c'), Literal('o'), Literal('v'), Literal('e'), Literal('r'), Literal('a'), Literal('g'), Literal('e'), Literal('.'), ZeroOrMore]) }
built glob set; 5 literals, 63 basenames, 10 extensions, 0 prefixes, 0 suffixes, 2 required extensions, 2 regexes
whitelisting python/deptry/config.py: Whitelist(IgnoreMatch(Types(Glob(Matched { def: FileTypeDef { name: "python", globs: ["*.py"] } }))))
whitelisting python/deptry/dependency_getter/pdm.py: Whitelist(IgnoreMatch(Types(Glob(Matched { def: FileTypeDef { name: "python", globs: ["*.py"] } }))))
...
Python files to scan for imports:
python/deptry/config.py
python/deptry/dependency_getter/pdm.py
python/deptry/dependency_getter/pep_621.py
```

Do we want that level of detail exposed to the user in the verbose logging?

**Review by fpgmaas — APPROVED** [review]

File: python/deptry/core.py (hunk 1 above)
Comment (mkniewallner) [reply python/deptry/core.py:17]: Yeah I must say that this is originally what I wanted to do, but it made tests more complex as they would need to instantiate `Core` class with a load of arguments we don't need to, as there is no default for the arguments. Eventually, I'd like to rework a bit the `Core` class, which right now feels like a god class with a too-generic name.

**Comment by mkniewallner** [comment]
> Looks good! I see that the verbose logging has become a bit more verbose:
> 
> ```
> Collecting Python files to scan...
> built glob set; 0 literals, 0 basenames, 0 extensions, 0 prefixes, 0 suffixes, 2 required extensions, 0 regexes
> glob converted to regex: Glob { glob: "**/*.py[cod]", re: "(?-u)^(?:/?|.*/)[^/]*\\.py[cod]$", opts: GlobOptions { case_insensitive: false, literal_separator: true, backslash_escape: true, empty_alternates: false }, tokens: Tokens([RecursivePrefix, ZeroOrMore, Literal('.'), Literal('p'), Literal('y'), Class { negated: false, ranges: [('c', 'c'), ('o', 'o'), ('d', 'd')] }]) }
> glob converted to regex: Glob { glob: "**/.coverage.*", re: "(?-u)^(?:/?|.*/)\\.coverage\\.[^/]*$", opts: GlobOptions { case_insensitive: false, literal_separator: true, backslash_escape: true, empty_alternates: false }, tokens: Tokens([RecursivePrefix, Literal('.'), Literal('c'), Literal('o'), Literal('v'), Literal('e'), Literal('r'), Literal('a'), Literal('g'), Literal('e'), Literal('.'), ZeroOrMore]) }
> built glob set; 5 literals, 63 basenames, 10 extensions, 0 prefixes, 0 suffixes, 2 required extensions, 2 regexes
> whitelisting python/deptry/config.py: Whitelist(IgnoreMatch(Types(Glob(Matched { def: FileTypeDef { name: "python", globs: ["*.py"] } }))))
> whitelisting python/deptry/dependency_getter/pdm.py: Whitelist(IgnoreMatch(Types(Glob(Matched { def: FileTypeDef { name: "python", globs: ["*.py"] } }))))
> ...
> Python files to scan for imports:
> python/deptry/config.py
> python/deptry/dependency_getter/pdm.py
> python/deptry/dependency_getter/pep_621.py
> ```
> 
> Do we want that level of detail exposed to the user in the verbose logging?

Unfortunately I don't think there's a way to control that, at least not easily. Setting `debug` log level mode in Python will also set the same level in Rust through `pyo3`, so both languages will output debug logs when enabling debug mode.

**Pull request merged**